### PR TITLE
Framework: Improve CSS management

### PR DESCRIPTION
**Pull request by Bontrey**
Empty PR as a forum for discussion while I get my bearings

Context: project suggested by @ehg 
Potential goals:
- Save on CSS boilerplate
- Download on-demand
- Play nicely with the idea of plugins
- ??? (suggestions welcomed)

### CODE REVIEW

**Comment by ehg** [comment]
I personally like the approach of picking up the .scss files from the directory structure, as it doesn't depart from the current way of doing things too much. Modifying the build process to account for this needs to take a few considerations into account:

- Speed. We don't want to slow down our already long build process.
- Ordering. There could be issues with dependent SCSS rules.
- Chunks. If we're going to split CSS into chunks, we need to consider whatever approach we take for picking up the SCSS files automatically, plays well with splitting the CSS into chunks ( in `webpack`, likely? )

@mtias may have some thoughts here.

**Comment by Bontrey** [comment]
I agree that any solution should definitely live on the assumption that the styles for a particular component live in the same directory.

There are some low-hanging fruits such as generating the master file of @imports, but I think we want to go farther.

I'm assuming that right now component js is split and asynchronously loaded based on routes only, but that eventually we would want to be able to do it on a per-component basis. And when we load the js for a component, we also want to load the styles. I'm researching the mechanism to do this right now

There is a lot of talk about inlining CSS into react components as JS, but I haven't seen any examples that looks friendly with SASS.

Been doing some reading on webpack, and there are some examples where a js file requires a css file. From my understanding, by default, the css is included in the output js. If an asynchronously loadable piece of js can simply place a <style> in the page header when it loads, then that should solve most of our problems.

Stay tuned..

**Comment by Bontrey** [comment]
Simple proof of concept for that last part I mentioned, loading CSS lazily with webpack:
https://bontrey.github.io/css-experiment/
https://github.com/Bontrey/css-experiment

Next is to figure out how to integrate this mechanism into the codebase in a generalized and non-intrusive way

**Comment by ehg** [comment]
> There are some low-hanging fruits such as generating the master file of @imports, but I think we want to go farther.

Yep. The `@import`s are frustrating, as you can see recently in https://github.com/Automattic/wp-calypso/issues/8001

> I'm assuming that right now component js is split and asynchronously loaded based on routes only, but that eventually we would want to be able to do it on a per-component basis. And when we load the js for a component, we also want to load the styles. I'm researching the mechanism to do this right now

Your assumption is basically right. We have https://github.com/Automattic/wp-calypso/pull/8544 in the works, which when it lands, we will start splitting up by components. 

> There is a lot of talk about inlining CSS into react components as JS, but I haven't seen any examples that looks friendly with SASS.

I don't think we will be doing any inlining anytime soon, as it would require a pretty huge change.

**Comment by ehg** [comment]
> Simple proof of concept for that last part I mentioned, loading CSS lazily with webpack

Looks good, I'm looking forward to the next step — so we can see early on its effect on build performance.

**Comment by mtias** [comment]
@ehg the ordering should be of no consequence. Every component needs to prefix to a global and unique namespace, and be very diligent with specificity when they want to overwrite rules under their namespace. If there are issues that come out of the ordering they need to be fixed.

**Comment by ehg** [comment]
> @ehg the ordering should be of no consequence. Every component needs to prefix to a global and unique namespace, and be very diligent with specificity when they want to overwrite rules under their namespace. If there are issues that come out of the ordering they need to be fixed.

Yep, I should've been clearer. Any issues we have with ordering, we do need to fix in the SASS. Thanks :)

**Comment by Bontrey** [comment]
Also plan to look into incorporating [CSS modules](https://github.com/css-modules/css-modules) so that people don't have to manually prefix

**Comment by Bontrey** [comment]
At first glance per-component styling seems to work pretty well with minimal effort

Some things to think about:
- Can there be flashes of unstyled content?
- Can there be conflicts from having two CSS pipelines?
- How to minimize need for imports of shared stuff in component stylesheets? Do we want to? (They are actual dependencies, after all)
- [Potential for duplication](https://github.com/jtangelder/sass-loader/issues/145)

**Comment by ehg** [comment]
> Can there be flashes of unstyled content?

As in, is that acceptable? In this case, no :)

> How to minimize need for imports of shared stuff in component stylesheets? Do we want to? (They are actual dependencies, after all)

Could we have a shared bundle?

**Comment by Bontrey** [comment]
> As in, is that acceptable? In this case, no :)

Nah, as in 'does it happen?' My sense is no, with webpack's runtime, but it's something to verify.

> Could we have a shared bundle?

If you mean a dependency that depends on all the shared stuff, probably. It remains to be seen whether this has a dramatic impact on performance.

File: client/post-editor/editor-sharing/accordion.jsx [hunk 1]
```diff
@@ -28,6 +28,8 @@ import { getSiteUserConnections } from 'state/sharing/publicize/selectors';
 import { postTypeSupports } from 'state/post-types/selectors';
 import { fetchConnections as requestConnections } from 'state/sharing/publicize/actions';
 
+require('./style.scss');
```
Comment (Bontrey) [inline client/post-editor/editor-sharing/accordion.jsx]: Not convinced that the saving of up to 1 line per jsx file is worth this magic.

Developers understand specifying module dependencies in code. Our problem right now is that the CSS dependencies are specified in a separate tree from JS (component) dependencies. If we solve that, I think having localized requires is fine.

if we want to use CSS Modules at some point, we'd also need the imported object

**Comment by lancewillett** [comment]
Closing for now; feel free to re-open in the future if we re-look at this again.

**Pull request closed without merging**